### PR TITLE
Limit # of child runs created for hyperparam search models

### PR DESCRIPTION
**Pull request by mohamad-arabi**
Signed-off-by: Mohamad Arabi <mohamad.arabi@databricks.com>

## What changes are proposed in this pull request?
The ability to limit the number of child runs created when running hyper param search with auto-logging. 
The default number of child runs is set to 5.  

## How is this patch tested?
Added pytest unit tests.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

A user can now specify a limit to the number child runs created when training an sklearn hyper parameter search estimator.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sklearn/__init__.py [hunk 1]
```diff
@@ -463,6 +463,7 @@ def autolog(
     exclusive=False,
     disable_for_unsupported_versions=False,
     silent=False,
+    max_hyper_param_runs=5,
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:466]:
```suggestion
    max_tuning_runs=5,
```

File: mlflow/sklearn/__init__.py [hunk 2]
```diff
@@ -654,6 +655,11 @@ def fetch_logged_data(run_id):
     :param silent: If ``True``, suppress all event logs and warnings from MLflow during scikit-learn
                    autologging. If ``False``, show all events and warnings during scikit-learn
                    autologging.
+    :param max_hyper_param_runs: The max number of child mlflow runs to be created for
+                                 hyper parameter search based estimators. If one only cares about
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:659]:
```suggestion
    :param max_hyper_param_runs: The maximum number of child MLflow runs to be created for
                                 hyperparameter search estimators. If one only cares about
```

File: mlflow/sklearn/__init__.py [hunk 3]
```diff
@@ -654,6 +655,11 @@ def fetch_logged_data(run_id):
...
+                                 creating an mlflow child run for the best k results from the
+                                 search, then set max_hyper_param_runs to k. The default value is
+                                 to track the best 5 hyper paramters.
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:662]: Can we refactor this phrasing to more succinctly tell the user that the best child runs are selected and also tell them how these runs are selected (i.e. via `rank_test_score` or the first available custom metric)?

File: mlflow/sklearn/__init__.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/sklearn/__init__.py:659]: Can we also add documentation indicating that the user can specify `None` to omit a maximum limit and log a child run for each set of explored hyperparameters?

File: mlflow/sklearn/utils.py [hunk 4]
```diff
@@ -615,9 +624,19 @@ def _create_child_runs_for_parameter_search(cv_estimator, parent_run, child_tags
...
     cv_results_df = pd.DataFrame.from_dict(cv_estimator.cv_results_)
-    for _, result_row in cv_results_df.iterrows():
+    rank_column_name = "rank_test_score"
+    if rank_column_name not in cv_results_df.columns.values:
+        rank_column_name = first_custom_rank_column(cv_results_df)
+        warnings.warn(
+            "Top {} child runs will be created based on ordering in {} column. ".format(
+                max_hyper_param_runs, rank_column_name,
+            )
+            + "One can choose not to limit the number of child runs created if so desired."
```
Comment (dbczumar) [inline mlflow/sklearn/utils.py:635]: How can users do this? I don't see where this PR provides that capability. I'd propose introducing `None` as a valid parameter value to satisfy this use case; we can check for `None` here and avoid limiting results.

Comment (dbczumar) [inline mlflow/sklearn/utils.py:635]: Can we address the user in second person (`you`) and tell them specifically how to remove the child runs limit (i.e. by passing `None` for the autologging call option)? Can we also tell them that the metric being selected is the first one found & that they can reorder their custom metrics to in order to choose a different one for child run creation?

File: mlflow/sklearn/utils.py [hunk 5]
```diff
@@ -597,6 +600,12 @@ def _create_child_runs_for_parameter_search(cv_estimator, parent_run, child_tags
     """
     import pandas as pd
 
+    def first_custom_rank_column(df):
+        column_names = df.columns.values
+        for i, col_name in enumerate(column_names):
+            if "rank_test_" in col_name:
+                return column_names[i]
```
Comment (dbczumar) [inline mlflow/sklearn/utils.py:607]: Nit: Do we need the index `i` here? Can we do `for col_name in column_names` instead?

File: mlflow/sklearn/utils.py [hunk 6]
```diff
@@ -597,6 +600,12 @@ def _create_child_runs_for_parameter_search(cv_estimator, parent_run, child_tags
     """
     import pandas as pd
 
+    def first_custom_rank_column(df):
```
Comment (dbczumar) [inline mlflow/sklearn/utils.py:603]: What happens if  a custom rank column is not identified? Perhaps we should pick the first `n` runs and warn the user?

Comment (mohamad-arabi) [reply mlflow/sklearn/utils.py:603]: It should be identified, that is what we expect from [cv_results_](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV). If a rank column is not identified then I think something bigger has gone wrong in the training and this will be gracefully handled because all of [_create_child_runs_for_parameter_search](https://github.com/mlflow/mlflow/blob/master/mlflow/sklearn/__init__.py#L819) is wrapped in a try catch

File: mlflow/sklearn/__init__.py [hunk 7]
```diff
@@ -654,6 +655,17 @@ def fetch_logged_data(run_id):
     :param silent: If ``True``, suppress all event logs and warnings from MLflow during scikit-learn
                    autologging. If ``False``, show all events and warnings during scikit-learn
                    autologging.
+    :param max_tuning_runs: The maximum number of child Mlflow runs to be created for
+                            hyperparameter search estimators. If one only cares about
+                            creating an mlflow child run for the best k results(based on
+                            `rank_test_score` values) from the search, then set max_tuning_runs
+                            to k. The default value is to track the best 5 search parameter sets.
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:662]: As discussed offline, can we say something like

```
To create child runs for the best `k` results from the search, set `max_tuning_runs` to `k`.
```

instead?

File: mlflow/sklearn/__init__.py [hunk 8]
```diff
@@ -654,6 +655,17 @@ def fetch_logged_data(run_id):
     :param silent: If ``True``, suppress all event logs and warnings from MLflow during scikit-learn
                    autologging. If ``False``, show all events and warnings during scikit-learn
                    autologging.
+    :param max_tuning_runs: The maximum number of child Mlflow runs to be created for
+                            hyperparameter search estimators. If one only cares about
+                            creating an mlflow child run for the best k results(based on
+                            `rank_test_score` values) from the search, then set max_tuning_runs
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:661]: Can we remove (`based on rank_test_score values`) and move this discussion about how the best values are selected to the bottom of the docstring along with the information about custom scorers? These two are closely related.

File: mlflow/sklearn/utils.py (hunk 6 above)
Comment (dbczumar) [reply mlflow/sklearn/utils.py:603]: I don't see `rank_test` explicitly defined in the contract of [cv_results_](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV), though it's used in an example.

Regardless, try/catch is fine as long as we have test coverage here. Can we confirm that we test our handling of child run creation failures for hyperparameter search and introduce coverage if we don't?

File: mlflow/sklearn/utils.py [hunk 9]
```diff
@@ -615,9 +624,24 @@ def _create_child_runs_for_parameter_search(cv_estimator, parent_run, child_tags
...
+    else:
+        rank_column_name = "rank_test_score"
+        if rank_column_name not in cv_results_df.columns.values:
+            rank_column_name = first_custom_rank_column(cv_results_df)
+            warnings.warn(
+                "Top {} child runs will be created based on ordering in {} column. ".format(
+                    max_tuning_runs, rank_column_name,
+                )
+                + "You can choose not to limit the number of child runs created by"
+                + "setting `max_tuning_runs=None`."
```
Comment (dbczumar) [inline mlflow/sklearn/utils.py:640]: Nit:

```suggestion
                + " setting `max_tuning_runs=None`."
```

File: mlflow/sklearn/utils.py (hunk 6 above)
Comment (mohamad-arabi) [reply mlflow/sklearn/utils.py:603]: Screen Shot 2021-05-24 at 12 29 37 PM

The example is the only location where they provide docs for `cv_results_`

Comment (mohamad-arabi) [reply mlflow/sklearn/utils.py:603]: Also we already do have test coverage for graceful failing of [_create_child_runs_for_parameter_search()](https://github.com/mlflow/mlflow/blob/master/tests/sklearn/test_sklearn_autolog.py#L848)

Comment (dbczumar) [reply mlflow/sklearn/utils.py:603]: Not to be pedantic, but the above text is just an example indicating how a Pandas DataFrame with particular columns (e.g. `rank_test_score`) is translated into a `cv_results_` dictionary; it does not explicitly stipulate that `rank_test_*` metrics are always computed, even if that is the case. Regardless, glad we have coverage! Thank you for auditing!

Comment (mohamad-arabi) [reply mlflow/sklearn/utils.py:603]: No problem, thanks for the diligent review of my PR!

Comment (mohamad-arabi) [reply mlflow/sklearn/utils.py:603]: That makes sense, they never explicitly listed what are the contents of `cv_results_` technically.

File: mlflow/sklearn/utils.py [hunk 10]
```diff
@@ -615,9 +624,30 @@ def _create_child_runs_for_parameter_search(cv_estimator, parent_run, child_tags
...
+                + " You can choose not to limit the number of child runs created by"
+                + " setting `max_tuning_runs=None`."
+            )
+        cv_results_best_n_df = cv_results_df.nsmallest(max_tuning_runs, rank_column_name)
+        # Log how many child runs will be created vs omitted.
+        rest = len(cv_results_df) - max_tuning_runs
+        if rest > 0:
+            _logger.info(
+                "Logging the %s best runs, %s others will be omitted.", max_tuning_runs, rest,
+            )
```
Comment (dbczumar) [inline mlflow/sklearn/utils.py:648]: Nit: This message becomes a bit odd when `max_tuning_runs` is zero or one.

"Logging the best 0 runs, 15 others will be omitted."

"Logging the best 1 runs, 14 others will be omitted."

Can we handle these cases?

File: mlflow/sklearn/__init__.py [hunk 11]
```diff
@@ -682,6 +693,11 @@ def fetch_logged_data(run_id):
         MAX_ENTITY_KEY_LENGTH,
     )
 
+    if max_tuning_runs is not None and max_tuning_runs < 0:
+        raise ValueError(
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:697]: Can we raise `MlflowException` with error code `INVALID_PARAMETER_VALUE` instead?

**Review by dbczumar — APPROVED** [review]
@mohamad-arabi LGTM once two comments have been addressed. Awesome work!
